x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.process;
+
+import org.elasticsearch.ResourceNotFoundException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.LocalNodeMasterListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.persistent.PersistentTasksCustomMetaData;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.ml.MlMetadata;
+import org.elasticsearch.xpack.core.ml.MlTasks;
+import org.elasticsearch.xpack.core.ml.action.OpenJobAction;
+import org.elasticsearch.xpack.core.ml.job.config.Job;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.job.JobManager;
+import org.elasticsearch.xpack.ml.job.persistence.JobResultsProvider;
+
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * This class keeps track of the memory requirement of ML jobs.
+ * It only functions on the master node - for this reason it should only be used by master node actions.
+ * The memory requirement for open ML jobs is updated at the following times:
+ * 1. When a master node is elected the memory requirement for all non-closed ML jobs is updated
+ * 2. The memory requirement for all non-closed ML jobs is updated periodically thereafter - every 30 seconds by default
+ * 3. When a job is opened the memory requirement for that single job is updated
+ * As a result of this every open job should have a value for its memory requirement that is no more than 30 seconds out-of-date.
+ */
+public class MlMemoryTracker extends AbstractComponent implements LocalNodeMasterListener {
+
+    public static final Setting<TimeValue> ML_MEMORY_UPDATE_FREQUENCY =
```

Comment:
This is a high update rate given that the latest memory usage is required for just 2 situations 1. opening a job and 2. re-assigning a job when it re-locates.

If opening a job refreshed the memory usage for all jobs that would solve 1. For 2. can this class listen for cluster events such as an ml node joining or leaving the cluster or a job task becoming unassigned and refresh the memory then. This would create a race between refreshing the memory and making the assignment decision so we would need a way to delay assignment. (Ask the MlMemoryTracker if it has been refreshed recently?). If 2. cannot be addressed then consider raising the update interval as we are handling the case where a job becomes unassigned and that should happen infrequently, a 10 minute update frequency may be more appropriate.

Reply:
Yes, true, it's the reallocation that causes the grief.  If we were allowed to do searches in the allocation method then the problem would be solved.  I don't think we can tolerate a 10 minute allocation interval, as job memory can change a lot when jobs first start up.  But you're right it should be possible to temporarily leave jobs unallocated while a refresh takes place, then trigger reallocation by removing that special reason from the persistent task.